### PR TITLE
use heartbeat packet for zero-window probe

### PR DESCRIPTION
**Pull request by jthomas43**
Uses a heartbeat packet instead of a data-carrying packet for zero-window probe.
Also fixes two bugs related to rwnd:
- fixes issue where a write queued to an idle stream with 0 send_rwnd would not start the ZWP timer
- fixes issue where a stream not send or zero-window probe if the rwnd fell to less than a full packet but > 0.

### CODE REVIEW

File: test/stream-write-read-receive-window-2.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  udx_stream_set_rwnd_max(&recv_stream, APPLICATION_BUFFER_SIZE);
+  recv_stream.get_read_buffer_size = application_get_buffer;
+}
+
+void
+application_on_read (udx_stream_t *stream, ssize_t read_len, const uv_buf_t *buf) {
+
+  printf("read=%zd", read_len);
+  if (read_len > 0) {
+    application.nbytes_in_buffer += read_len;
```
Comment (marcus-pousette-hp) [inline test/stream-write-read-receive-window-2.c:68]: If APPLICATION_BUFFER_SIZE is a hard limit then this should also pass, but seems not

```suggestion
    application.nbytes_in_buffer += read_len;
    assert(application.nbytes_in_buffer <= APPLICATION_BUFFER_SIZE);
```

File: src/udx.c [hunk 2]
```diff
@@ -2675,8 +2673,9 @@ _udx_stream_write (udx_stream_write_t *write, udx_stream_t *stream, const uv_buf
   }
 
   // if an idle, zero window stream has data queued, send a zero-window probe immediately
-  if (stream->writes_queued_bytes > 0 && stream->send_rwnd == 0) {
-    send_new_packet(stream, UDX_PROBE_TYPE_ZWP);
+  if (stream->send_rwnd == 0) {
+    send_probe(stream);
+    stream_timer_start(stream, UDX_TIMER_ZWP, stream->rto);
```
Comment (marcus-pousette-hp) [inline src/udx.c:2686]: Could we add this regression test, since receiving a zero-window update and then queueing another write currently replaces the active RTO/TLP with ZWP, leaving an older dropped packet unretransmitted and both writes stalled?


Below is a test (ai-generated). I would assume should pass 


```ts
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "../include/udx.h"

/*
 * An older data packet is lost before the peer advertises a zero receive
 * window. Queueing another write must not let ZWP displace recovery of that
 * older packet. Once it is retransmitted, the receiver consumes it, opens its
 * window, and both writes can complete.
 */

#define FIRST_DATA_SIZE  (UDX_MTU_BASE - UDX_IPV4_HEADER_SIZE)
#define SECOND_DATA_SIZE 1
#define TEST_TIMEOUT_MS  2000

uv_loop_t loop;
udx_t udx;

udx_socket_t recv_sock;
udx_stream_t recv_stream;

udx_socket_t send_sock;
udx_stream_t send_stream;

udx_stream_write_t *first_req;
udx_stream_write_t *second_req;
udx_stream_send_t window_update_req;

uv_check_t observe_zero_window;
uv_timer_t timeout;

char first_data[FIRST_DATA_SIZE];
char second_data[SECOND_DATA_SIZE];

bool second_write_queued;
int writes_acked;
size_t bytes_read;

uint32_t
pretend_buffer_is_full (udx_stream_t *stream) {
  return stream->recv_rwnd_max;
}

uint32_t
pretend_buffer_is_empty (udx_stream_t *stream) {
  return 0;
}

void
on_read (udx_stream_t *stream, ssize_t read_len, const uv_buf_t *buf) {
  assert(read_len == (ssize_t) buf->len);
  bytes_read += read_len;

  // Receiving the missing packet lets the application free its buffer. The
  // ACK for this packet will therefore advertise an open receive window.
  recv_stream.get_read_buffer_size = &pretend_buffer_is_empty;
}

void
on_ack (udx_stream_write_t *req, int status, int unordered) {
  assert(status == 0);
  writes_acked++;

  if (writes_acked == 2) {
    uv_stop(&loop);
  }
}

void
on_observe_zero_window (uv_check_t *check) {
  if (second_write_queued || send_stream.send_rwnd != 0) {
    return;
  }

  second_write_queued = true;
  uv_check_stop(check);

  uv_buf_t buf = uv_buf_init(second_data, sizeof(second_data));
  int e = udx_stream_write(
    second_req,
    &send_stream,
    &buf,
    1,
    on_ack
  );
  assert(e && "drained");
}

void
on_timeout (uv_timer_t *timer) {
  uv_stop(timer->loop);
}

void
bind_addr (struct sockaddr_in *addr, int port) {
  int e = uv_ip4_addr("127.0.0.1", port, addr);
  assert(e == 0);
}

int
main () {
  int e;

  first_req = malloc(udx_stream_write_sizeof(1));
  second_req = malloc(udx_stream_write_sizeof(1));

  e = uv_loop_init(&loop);
  assert(e == 0);

  e = udx_init(&loop, &udx, NULL);
  assert(e == 0);

  struct sockaddr_in recv_addr;
  struct sockaddr_in send_addr;
  bind_addr(&recv_addr, 9131);
  bind_addr(&send_addr, 9132);

  e = udx_socket_init(&udx, &recv_sock, NULL);
  assert(e == 0);
  e = udx_socket_bind(
    &recv_sock,
    (struct sockaddr *) &recv_addr,
    0
  );
  assert(e == 0);

  e = udx_socket_init(&udx, &send_sock, NULL);
  assert(e == 0);
  e = udx_socket_bind(
    &send_sock,
    (struct sockaddr *) &send_addr,
    0
  );
  assert(e == 0);

  e = udx_stream_init(&udx, &recv_stream, 1, NULL, NULL);
  assert(e == 0);
  e = udx_stream_init(&udx, &send_stream, 2, NULL, NULL);
  assert(e == 0);

  recv_stream.get_read_buffer_size = &pretend_buffer_is_full;
  send_stream.send_rwnd = FIRST_DATA_SIZE;
  send_stream.rto = 200;

  e = udx_stream_connect(
    &recv_stream,
    &recv_sock,
    2,
    (struct sockaddr *) &send_addr
  );
  assert(e == 0);

  e = udx_stream_connect(
    &send_stream,
    &send_sock,
    1,
    (struct sockaddr *) &recv_addr
  );
  assert(e == 0);

  e = udx_stream_read_start(&recv_stream, on_read);
  assert(e == 0);

  memset(first_data, 0, sizeof(first_data));
  memset(second_data, 0, sizeof(second_data));

  // Send one full packet and drop only its first transmission.
  udx.debug_flags |= UDX_DEBUG_FORCE_DROP_DATA;

  uv_buf_t first_buf = uv_buf_init(
    first_data,
    sizeof(first_data)
  );

  e = udx_stream_write(
    first_req,
    &send_stream,
    &first_buf,
    1,
    on_ack
  );
  assert(e && "drained");

  udx.debug_flags &= ~UDX_DEBUG_FORCE_DROP_DATA;

  assert(send_stream.remote_acked != send_stream.seq);
  assert(
    send_stream.pending_timer == UDX_TIMER_RTO ||
    send_stream.pending_timer == UDX_TIMER_TLP
  );

  // Advertise zero window without advancing the cumulative ACK.
  uv_buf_t update = uv_buf_init("", 0);

  e = udx_stream_send(
    &window_update_req,
    &recv_stream,
    &update,
    1,
    NULL
  );
  assert(e == 0);

  // Queue the second write only after the sender observes zero window.
  e = uv_check_init(&loop, &observe_zero_window);
  assert(e == 0);

  e = uv_check_start(
    &observe_zero_window,
    on_observe_zero_window
  );
  assert(e == 0);

  e = uv_timer_init(&loop, &timeout);
  assert(e == 0);

  e = uv_timer_start(
    &timeout,
    on_timeout,
    TEST_TIMEOUT_MS,
    0
  );
  assert(e == 0);

  uv_run(&loop, UV_RUN_DEFAULT);

  assert(second_write_queued);

  assert(
    send_stream.retransmit_count > 0 &&
    "zero-window state displaced loss recovery"
  );

  assert(
    bytes_read == FIRST_DATA_SIZE + SECOND_DATA_SIZE
  );

  assert(writes_acked == 2);

  free(first_req);
  free(second_req);

  return 0;
}
```

Comment (marcus-pousette-hp) [inline src/udx.c:2686]: Is it intentional that each write during an existing zero-window episode sends another immediate probe and restarts the ZWP deadline, or should probing remain timer-driven after the initial probe?

See test below that is currently failing (assuming a more stricter behavior is wanted). Disclaimer AI-generated

```ts

#include <assert.h>
#include <stdlib.h>

#include "../include/udx.h"

/* Repeated writes in one zero-window episode neither probe nor rearm again. */

#define NWRITES       3
#define RTO_MS        500
#define QUEUE_MORE_MS 100

uv_loop_t loop;
udx_t udx;

udx_socket_t recv_sock;
udx_stream_t recv_stream;

udx_socket_t send_sock;
udx_stream_t send_stream;

udx_stream_write_t *reqs[NWRITES];
char data[NWRITES];

uv_timer_t queue_more;

uint64_t packets_tx_before;
uint64_t due_before;
uint64_t due_after;

uint32_t
pretend_buffer_is_full (udx_stream_t *stream) {
  return stream->recv_rwnd_max;
}

void
bind_addr (struct sockaddr_in *addr, int port) {
  int e = uv_ip4_addr("127.0.0.1", port, addr);
  assert(e == 0);
}

void
on_queue_more (uv_timer_t *timer) {
  assert(send_stream.pending_timer == UDX_TIMER_ZWP);
  due_before = uv_timer_get_due_in(&send_stream.timer);

  for (int i = 1; i < NWRITES; i++) {
    reqs[i] = malloc(udx_stream_write_sizeof(1));
    uv_buf_t buf = uv_buf_init(&data[i], 1);

    int e = udx_stream_write(
      reqs[i],
      &send_stream,
      &buf,
      1,
      NULL
    );
    assert(e && "drained");
  }

  due_after = uv_timer_get_due_in(&send_stream.timer);
  uv_stop(timer->loop);
}

int
main () {
  int e;

  e = uv_loop_init(&loop);
  assert(e == 0);

  e = udx_init(&loop, &udx, NULL);
  assert(e == 0);

  struct sockaddr_in recv_addr;
  struct sockaddr_in send_addr;

  bind_addr(&recv_addr, 9151);
  bind_addr(&send_addr, 9152);

  e = udx_socket_init(&udx, &recv_sock, NULL);
  assert(e == 0);

  e = udx_socket_bind(
    &recv_sock,
    (struct sockaddr *) &recv_addr,
    0
  );
  assert(e == 0);

  e = udx_socket_init(&udx, &send_sock, NULL);
  assert(e == 0);

  e = udx_socket_bind(
    &send_sock,
    (struct sockaddr *) &send_addr,
    0
  );
  assert(e == 0);

  e = udx_stream_init(
    &udx,
    &recv_stream,
    1,
    NULL,
    NULL
  );
  assert(e == 0);

  e = udx_stream_init(
    &udx,
    &send_stream,
    2,
    NULL,
    NULL
  );
  assert(e == 0);

  recv_stream.get_read_buffer_size = &pretend_buffer_is_full;
  send_stream.send_rwnd = 0;
  send_stream.rto = RTO_MS;

  e = udx_stream_connect(
    &recv_stream,
    &recv_sock,
    2,
    (struct sockaddr *) &send_addr
  );
  assert(e == 0);

  e = udx_stream_connect(
    &send_stream,
    &send_sock,
    1,
    (struct sockaddr *) &recv_addr
  );
  assert(e == 0);

  packets_tx_before = send_stream.packets_tx;

  reqs[0] = malloc(udx_stream_write_sizeof(1));
  uv_buf_t buf = uv_buf_init(&data[0], 1);

  e = udx_stream_write(
    reqs[0],
    &send_stream,
    &buf,
    1,
    NULL
  );
  assert(e && "drained");

  e = uv_timer_init(&loop, &queue_more);
  assert(e == 0);

  e = uv_timer_start(
    &queue_more,
    on_queue_more,
    QUEUE_MORE_MS,
    0
  );
  assert(e == 0);

  uv_run(&loop, UV_RUN_DEFAULT);

  assert(send_stream.writes_queued_bytes == NWRITES);
  assert(send_stream.pending_timer == UDX_TIMER_ZWP);

  assert(
    due_after <= due_before &&
    "a later write restarted the existing ZWP deadline"
  );

  assert(
    send_stream.packets_tx - packets_tx_before == 1 &&
    "each zero-window write sent another immediate probe"
  );

  return 0;
}

```

**Review by marcus-pousette-hp — CHANGES_REQUESTED** [review]
this is nice stuff! I have some comments, after resolved and tests are passing, I am interested later to do a review of boths PRs aside of each other for perf

File: test/stream-write-read-receive-window-2.c (hunk 1 above)
Comment (jthomas43) [reply test/stream-write-read-receive-window-2.c:68]: It is not a hard limit as of now. It is basically a tunable number that gives backpressure based on how many bytes are in some (maybe virtual) application buffer. If the application buffer is full (or over full) then set an rwnd of 0. The most egregious overflow is from out-of-order packets, which are buffered by libudx and are not visible to the application until the gap packet arrives. Tail loss probe is another case where we can push past the window.

File: src/udx.c (hunk 2 above)
Comment (jthomas43) [reply src/udx.c:2686]: No it should only start if the write causes a transition from idle -> having data or an END packet queued. Fixing

File: test/stream-write-read-receive-window-2.c (hunk 1 above)
Comment (marcus-pousette-hp) [reply test/stream-write-read-receive-window-2.c:68]: Okay good then I am happy

File: src/udx.c (hunk 2 above)
Comment (jthomas43) [reply src/udx.c:2686]: I'll take a look, in general we should not set a ZWP if data is in flight. In TCP a receiver shouldn't send a zero-window probe with data in flight since it would be disallowing data it previously allowed. I'm deciding between being robust against a zero-window with data in flight, or tightening the rwnd semantics to make sure it can't happen.

Comment (jthomas43) [reply src/udx.c:2686]: - fixed:  will only arm a ZWP timer when new data is queued on an idle connection. 
- I added the test. Besides testing that the ZWP timer is not armed when queuing data on an already transmitting connection, this test also sends a zero window update while data it previously permitted is in flight. I allowed retransmits to ignore the rwnd to cope with this scenario.

File: test/stream-write-read-receive-window-2.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+  uv_run(&loop, UV_RUN_DEFAULT);
+
+  // zwp_count only counts the number of zwp timeouts, since we start with a zero window
+  // we automatically probe (without waiting for a timeout) when data is queued.
+
+  assert(send_stream.zwp_count > 10);
+  assert(recv_stream.zwp_count == 0);
+  assert(ack_called);
+  assert(send_stream.retransmit_count == 0);
```
Comment (marcus-pousette-hp) [inline test/stream-write-read-receive-window-2.c]: Would this revert this behavior? #300

File: src/udx.c [hunk 4]
```diff
@@ -2675,8 +2681,9 @@ _udx_stream_write (udx_stream_write_t *write, udx_stream_t *stream, const uv_buf
   }
 
   // if an idle, zero window stream has data queued, send a zero-window probe immediately
-  if (stream->writes_queued_bytes > 0 && stream->send_rwnd == 0) {
-    send_new_packet(stream, UDX_PROBE_TYPE_ZWP);
+  if (stream_was_idle && stream->send_rwnd == 0 && stream_was_idle) {
```
Comment (marcus-pousette-hp) [inline src/udx.c]: nit: duplicated stream_was_idle check

File: src/udx.c [hunk 5]
```diff
@@ -1184,14 +1185,16 @@ udx_rack_reo_timeout (uv_timer_t *timer) {
...
   udx_stream_t *stream = timer->data;
-  stream->pending_timer = UDX_TIMER_NONE;
   assert(stream->status & UDX_STREAM_CONNECTED);
   assert(stream->send_rwnd == 0);
   assert((stream->status & UDX_STREAM_CLOSED) == 0);
 
   stream->zwp_count++;
   debug_printf("zwp: stream=%u\n", stream->remote_id);
-  send_new_packet(stream, UDX_PROBE_TYPE_ZWP);
+  send_probe(stream);
```
Comment (marcus-pousette-hp) [inline src/udx.c:1194]: Sanity check: #272 introduced UDX_HEADER_HEARTBEAT and made upgraded peers ACK it, but data flow liveness did not depend on that ACK. However, this PR uses HEARTBEAT for ZWP, so a sender with send_rwnd == 0 now depends on the peer supporting #272’s ACK behavior. udx-native@1.20.1 has included that behavior for a while, so the operational impact may be limited, but should we explicitly document udx-native@1.20.1+ as the minimum compatible peer version?

**Review by marcus-pousette-hp — CHANGES_REQUESTED** [review]
We are getting close! and I think I start to like the probing solution rather than my alternative PR more in terms of it being less complicated. I am curios later if we can figure out some actual recovery and perf benches around this to get some number just how well it performs. 

I made 4 comments, to address


Co-reviewed with AI

File: src/udx.c [hunk 6]
```diff
@@ -687,12 +684,18 @@ send_ack (udx_stream_t *stream) {
...
+stream_may_send (udx_stream_t *stream, bool retransmit) {
   update_pacing_time(stream);
   if (stream->tb_available == 0) {
     return false;
   }
-  return stream->inflight_queue.len < send_window_in_packets(stream);
+  if (retransmit) {
+    // ignore rwnd for retransmits to cope with receiver shrinking the window
+    // while data is in flight.
+    return stream->inflight_queue.len < stream->cwnd;
```
Comment (marcus-pousette-hp) [inline src/udx.c:695]: So as you mentioend this a divergence from LINUX "oldest-only" behaviour, The current test drops only remote_acked, so it would pass under either implementation. If we make the test explicitly also allow other packets we can use that as documentation too for future changes

File: src/udx.c (hunk 5 above)
Comment (jthomas43) [reply src/udx.c:1194]: I don't think we have to do that until we expose `udx_stream_set_rwnd_max()` and `get_read_buffer_size()` in udx-native. Until then they have no way to send a zero-window update, trigger a zwp timer and probe, etc, so no reason to force an update.

File: test/stream-write-read-receive-window-2.c (hunk 3 above)
Comment (jthomas43) [reply test/stream-write-read-receive-window-2.c]: I'm not 100% sure, the test is a little different but it might have the same flake with `send_stream.retransmit_count == 0` failing. I don't remember the details, i think the TLP could fire quickly because of the 0rtt ?

I relaxed it from `stream->retransmit_count == 0` to `stream->rto_count == 0`. I think it's safe to say this test shouldn't have a full RTO.

File: src/udx.c (hunk 6 above)
Comment (jthomas43) [reply src/udx.c:695]: Are you suggesting we assert that we transmit the entire retransmit queue? If so, I think we should make a separate PR for that using the technique from #301

Comment (marcus-pousette-hp) [reply src/udx.c:695]: That was expressed poorly by me, what i meant was test at least one retransmission beyond remote_acked  , because that would document the divergence from Linux SND.UNA only behaviour

Comment (marcus-pousette-hp) [reply src/udx.c:695]: Doing that in a separate PR is fine too!

File: src/udx.c (hunk 5 above)
Comment (marcus-pousette-hp) [reply src/udx.c:1194]: uggh yeah, my bad. Forgot this. That makes very much sense. Then the assumption still is that there are no other libudx users using this lib directly. In a search on github i did find users of libudx directly but no consumers of these methods

File: test/stream-write-read-receive-window-2.c (hunk 3 above)
Comment (marcus-pousette-hp) [reply test/stream-write-read-receive-window-2.c]: I think the TLP can fire quickly because the loopback test establishes a very small nonzero srtt, and that would flake sometimes in the Github runner ref other PR https://github.com/holepunchto/libudx/pull/300

File: src/udx.c (hunk 6 above)
Comment (jthomas43) [reply src/udx.c:695]: added an [issue](https://github.com/holepunchto/libudx/issues/303 ) to track it.

File: src/udx.c [hunk 7]
```diff
@@ -1184,14 +1185,16 @@ udx_rack_reo_timeout (uv_timer_t *timer) {
...
   assert(stream->send_rwnd == 0);
   assert((stream->status & UDX_STREAM_CLOSED) == 0);
 
   stream->zwp_count++;
   debug_printf("zwp: stream=%u\n", stream->remote_id);
-  send_new_packet(stream, UDX_PROBE_TYPE_ZWP);
+  send_probe(stream);
+  assert(stream->pending_timer == UDX_TIMER_ZWP);
+  assert(stream->writes_queued_bytes > 0 || (stream->status & (UDX_STREAM_ENDING | UDX_STREAM_ENDED)) == UDX_STREAM_ENDING);
+  stream_timer_start(stream, UDX_TIMER_ZWP, stream->rto * 2);
```
Comment (marcus-pousette-hp) [inline src/udx.c:1197]: `stream->rto * 2` vs `stream->rto` 

why is rto * 2 used here. We might have talked about this before but I am not finding it

File: test/stream-no-zwp-with-data-in-flight.c [hunk 8]
```diff
@@ -0,0 +1,247 @@
+#include <assert.h>
```
Comment (marcus-pousette-hp) [inline test/stream-no-zwp-with-data-in-flight.c:1]: "stream-no-zwp-with-data-in-flight" 

Does this test actually prove that no ZWP is armed while data is in flight? It currently only asserts eventual retransmission and completion, and it also passes against #295

**Review by marcus-pousette-hp — DISMISSED** [review]
Another round, but smaller points now mainly related to tests. I did some perf analysis against #295 and this looks very promising and on par, the only question is my rto * 2 comment that perhaps could be questioned/evaluated more

I think the remaining explicit coverage gap may be a lost window opening response. The existing receive window tests cover repeated ZWP timeouts, but they always deliver the response that first advertises rwnd > 0. Would it be worth dropping that first positive rwnd response and asserting that a later ZWP is sent and the queued write completes?

File: src/udx.c (hunk 7 above)
Comment (jthomas43) [reply src/udx.c:1197]: A guess really: It was probably copied from the RTO's logic. RTO and ZWP should have exponential backoff according to the RFCs, instead we only only back off once.

File: test/stream-no-zwp-with-data-in-flight.c (hunk 8 above)
Comment (jthomas43) [reply test/stream-no-zwp-with-data-in-flight.c:1]: Sort of. There are two ways to enter ZWP, 
1. on receiving an ACK with: `delivered > 0 && remote_acked == seq && send_rwnd == 0 && (bytes_queued > 0 || end_queued)`
2. on queueing new data when `stream_was_idle && send_rwnd == 0`.

the test checks (2) with:
```
  assert(
    send_stream.retransmit_count > 0 &&
    "zero-window state displaced loss recovery"
  ); 
```

when the test was generated it failed because (2) didn't have the same `stream_was_idle` check, so we could enter ZWP instead of staying in RTO and would not retransmit data.

So it doesn't categorically prove that it never enters ZWP but it proves that we don't enter ZWP with this specific code through this particular path. I think the name could be better maybe but it's OK with the comment:

```
/* 
 * An older data packet is lost before the peer advertises a zero receive 
 * window. Queueing another write must not let ZWP displace recovery of that 
 * older packet. Once it is retransmitted, the receiver consumes it, opens its 
 * window, and both writes can complete. 
 */ 
 ```

File: docs/wireshark/udx.lua [hunk 9]
```diff
@@ -81,17 +88,21 @@ function udx.dissector(tvb, pinfo, tree)
...
+        len_str = " Len=" .. (len-pos)
     end
 
     local id = tvb(4,4):le_uint()
+    local wnd = tvb(8,4):le_uint()
     local seq = tvb(12,4):le_uint()
     local ack = tvb(16,4):le_uint()
-    local info = pinfo.src_port .. " → " .. pinfo.dst_port ..
-        " Id=" .. id ..
+    local info = " Id=" .. id ..
```
Comment (lejeunerenard) [inline docs/wireshark/udx.lua:98]: Looks like a refactor / change unrelated to ZWP but the src & dst ports were removed. Just flagging to ensure that's intended.

File: include/udx.h
Comment (lejeunerenard) [inline include/udx.h:1]: Note that this is not related ZWP but clean up of the header.

**Review by lejeunerenard — APPROVED** [review]

File: docs/wireshark/udx.lua (hunk 9 above)
Comment (jthomas43) [reply docs/wireshark/udx.lua:98]: intended, they can always see source and destination port in their own columns and the info column was getting crowded.
